src/python/gem5/prebuilt/viper/cpu_cache_hierarchy.py

```diff
@@ -0,0 +1,273 @@
+    AbstractRubyCacheHierarchy,
+)
+from ...components.cachehierarchies.ruby.caches.viper.corepair_cache import (
+    CorePairCache,
+)
+from ...components.cachehierarchies.ruby.caches.viper.directory import (
+    ViperCPUDirectory,
+)
+from ...components.cachehierarchies.ruby.caches.viper.dma_controller import (
+    ViperCPUDMAController,
+)
+from ...prebuilt.viper.board import ViperBoard
+from ...utils.override import overrides
+from ...utils.requires import requires
+from .viper_network import SimplePt2Pt
+
+
+class ViperCPUCacheHierarchy(AbstractRubyCacheHierarchy):
+    """
+    The VIPER CPU cache hierarchy creates CPU-side Ruby caches and connects
+    the nodes using a simple point-to-point topology.
+    """
+
+    def __init__(
+        self,
+        l1d_size: str,
+        l1d_assoc: int,
+        l1i_size: str,
+        l1i_assoc: int,
+        l2_size: str,
+        l2_assoc: int,
+        l3_size: str,
+        l3_assoc: int,
+    ):
+        """
+        :param size: The size of each cache in the heirarchy.
+        :param assoc: The associativity of each cache.
+        :param device_dmas: Optional list of CPU connect device DMAs
+        """
+        super().__init__()
+
+        self._l1d_size = l1d_size
+        self._l1d_assoc = l1d_assoc
+        self._l1i_size = l1i_size
+        self._l1i_assoc = l1i_assoc
+        self._l2_size = l2_size
+        self._l2_assoc = l2_assoc
+        self._l3_size = l3_size
+        self._l3_assoc = l3_assoc
+
+        self.ruby_system = RubySystem()
+
+    @overrides(AbstractCacheHierarchy)
+    def incorporate_cache(self, board: ViperBoard) -> None:
+        requires(coherence_protocol_required=CoherenceProtocol.GPU_VIPER)
+
+        # Ruby networks for CPU
+        self.ruby_system.network = SimplePt2Pt(self.ruby_system)
+
+        # MOESI_AMD_Base uses 5 virtual networks.
+        self.ruby_system.number_of_virtual_networks = 5
+        self.ruby_system.network.number_of_virtual_networks = 5
+
+        # There is a single local list of all of the controllers to make it
+        # easier to connect everything to the CPU network. This can be
+        # customized depending on the topology/network requirements.
+        # Create one controller for each L1 cache (and the cache mem obj.)
+        # Create a single directory controller (Really the memory cntrl).
+        self._controllers = []
+
+        cores = board.get_processor().get_cores()
+        num_cores = len(cores)
+        for i in range(0, num_cores, 2):
+            cache = CorePairCache(
+                l1d_size=self._l1d_size,
+                l1d_assoc=self._l1d_assoc,
+                l1i_size=self._l1i_size,
+                l1i_assoc=self._l1i_assoc,
+                l2_size=self._l2_size,
+                l2_assoc=self._l2_assoc,
+                network=self.ruby_system.network,
+                cache_line_size=board.get_cache_line_size(),
```

Comment:
Doesn't Ruby only support 64?  If so, not sure if we need an assert here?

Reply:
I have successfully used this to run 64B cache lines on CPU side and 128B cache line size on GPU side. I know some folks have used up to 256B as well.